### We'll be investigating the 3 json files for data quality in this notebook

In [1]:
# importing libraries
import json
import datetime as dt
import pandas as pd

In [2]:
# Function to convert data from unstructured json to structured pandas dataframe
def jsonToDf(file_name):
    file_name = [json.loads(line) for line in open((file_name+'.json'), 'r')]
    file_name = pd.json_normalize(file_name)
    return file_name

#### Checking users.json

In [3]:
users = jsonToDf('users')
users

active         role signUpSource state                  _id.$oid  \
0      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
1      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
2      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
3      True     consumer        Email    WI  5ff1e1eacfcf6c399c274ae6   
4      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
..      ...          ...          ...   ...                       ...   
490    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
491    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
492    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
493    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
494    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   

     createdDate.$date  lastLogin.$date  
0        1609687444800     1.609688e+12  
1        1609687444800     1.609688e+12  
2        1609687444800     1.609688e+12  
3        1609687530554     1.609688e+12  
4        1609687444800     1.609688e+12  
..                 ...              ...  
490      1418998882381     1.614963e+12  
491      1418998882381     1.614963e+12  
492      1418998882381     1.614963e+12  
493      1418998882381     1.614963e+12  
494      1418998882381     1.614963e+12  

[495 rows x 7 columns]

In [4]:
# Writing a function to convert EPOCH date since we have to do this a few times
def dateConvert(df, field_to_convert):
    df[field_to_convert] = pd.to_datetime(df[field_to_convert], unit='ms')

In [5]:
dateConvert(users, 'createdDate.$date')
dateConvert(users, 'lastLogin.$date')
users

active         role signUpSource state                  _id.$oid  \
0      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
1      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
2      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
3      True     consumer        Email    WI  5ff1e1eacfcf6c399c274ae6   
4      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
..      ...          ...          ...   ...                       ...   
490    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
491    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
492    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
493    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
494    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   

          createdDate.$date               lastLogin.$date  
0   2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872  
1   2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872  
2   2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872  
3   2021-01-03 15:25:30.554 2021-01-03 15:25:30.596999936  
4   2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872  
..                      ...                           ...  
490 2014-12-19 14:21:22.381 2021-03-05 16:52:23.204000000  
491 2014-12-19 14:21:22.381 2021-03-05 16:52:23.204000000  
492 2014-12-19 14:21:22.381 2021-03-05 16:52:23.204000000  
493 2014-12-19 14:21:22.381 2021-03-05 16:52:23.204000000  
494 2014-12-19 14:21:22.381 2021-03-05 16:52:23.204000000  

[495 rows x 7 columns]

In [6]:
# Checking number of uniques in '_id.$oid'
users['_id.$oid'].nunique()

212

Looks like there are many duplicates in this data. Let's investigate further

In [7]:
users1 = users.drop_duplicates()
users1

active         role signUpSource state                  _id.$oid  \
0      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
3      True     consumer        Email    WI  5ff1e1eacfcf6c399c274ae6   
6      True     consumer        Email    WI  5ff1e1e8cfcf6c399c274ad9   
7      True     consumer        Email    WI  5ff1e1b7cfcf6c399c274a5a   
9      True     consumer        Email    WI  5ff1e1f1cfcf6c399c274b0b   
..      ...          ...          ...   ...                       ...   
435    True  fetch-staff        Email    NH  5fc961c3b8cfca11a077dd33   
455    True  fetch-staff        Email   NaN  5fa41775898c7a11a6bcef3e   
456    True  fetch-staff       Google    AL  5fa32b4d898c7a11a6bcebce   
462    True  fetch-staff          NaN    IL  5964eb07e4b03efd0c0f267b   
475    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   

          createdDate.$date               lastLogin.$date  
0   2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872  
3   2021-01-03 15:25:30.554 2021-01-03 15:25:30.596999936  
6   2021-01-03 15:25:28.354 2021-01-03 15:25:28.392000000  
7   2021-01-03 15:24:39.626 2021-01-03 15:24:39.664999936  
9   2021-01-03 15:25:37.564 2021-01-03 15:25:37.599000064  
..                      ...                           ...  
435 2020-12-03 22:08:03.936 2021-02-26 22:39:16.799000064  
455 2020-11-05 15:17:09.396 2021-03-04 16:02:02.025999872  
456 2020-11-04 22:29:33.309 2021-03-04 07:21:58.047000064  
462 2017-07-11 15:13:11.771 2021-03-04 19:07:49.769999872  
475 2014-12-19 14:21:22.381 2021-03-05 16:52:23.204000000  

[212 rows x 7 columns]

Dropping the duplicates gives us exactly 212 rows which is the same as the number of unique IDs. '_id.$oid' can be used as the primary key now

Items in the table where 'role' = 'fetch-staff' should also probably be removed in order to not mess with any final result

#### Checking brands.json

In [8]:
brands = jsonToDf('brands')
brands

barcode            category      categoryCode  \
0     511111019862              Baking            BAKING   
1     511111519928           Beverages         BEVERAGES   
2     511111819905              Baking            BAKING   
3     511111519874              Baking            BAKING   
4     511111319917      Candy & Sweets  CANDY_AND_SWEETS   
...            ...                 ...               ...   
1162  511111116752              Baking            BAKING   
1163  511111706328  Breakfast & Cereal               NaN   
1164  511111416173      Candy & Sweets  CANDY_AND_SWEETS   
1165  511111400608             Grocery               NaN   
1166  511111019930              Baking            BAKING   

                           name topBrand                  _id.$oid  \
0     test brand @1612366101024    False  601ac115be37ce2ead437551   
1                     Starbucks    False  601c5460be37ce2ead43755f   
2     test brand @1612366146176    False  601ac142be37ce2ead43755d   
3     test brand @1612366146051    False  601ac142be37ce2ead43755a   
4     test brand @1612366146827    False  601ac142be37ce2ead43755e   
...                         ...      ...                       ...   
1162  test brand @1601644365844      NaN  5f77274dbe37ce6b592e90c0   
1163       Dippin Dots® Cereal      NaN  5dc1fca91dda2c0ad7da64ae   
1164  test brand @1598639215217      NaN  5f494c6e04db711dd8fe87e7   
1165          LIPTON TEA Leaves    False  5a021611e4b00efe02b02a57   
1166  test brand @1613158231643    False  6026d757be37ce6369301468   

                  cpg.$id.$oid cpg.$ref                      brandCode  
0     601ac114be37ce2ead437550     Cogs                            NaN  
1     5332f5fbe4b03c9a25efd0ba     Cogs                      STARBUCKS  
2     601ac142be37ce2ead437559     Cogs  TEST BRANDCODE @1612366146176  
3     601ac142be37ce2ead437559     Cogs  TEST BRANDCODE @1612366146051  
4     5332fa12e4b03c9a25efd1e7     Cogs  TEST BRANDCODE @1612366146827  
...                        ...      ...                            ...  
1162  5f77274dbe37ce6b592e90bf     Cogs                            NaN  
1163  53e10d6368abd3c7065097cc     Cogs             DIPPIN DOTS CEREAL  
1164  5332fa12e4b03c9a25efd1e7     Cogs  TEST BRANDCODE @1598639215217  
1165  5332f5f6e4b03c9a25efd0b4     Cogs              LIPTON TEA Leaves  
1166  6026d757be37ce6369301467     Cogs  TEST BRANDCODE @1613158231644  

[1167 rows x 9 columns]

In [9]:
# Checking number of unique values in '_id.$oid'
brands['_id.$oid'].nunique()

1167

In [10]:
# Checking number of unique values in 'barcode'
brands['barcode'].nunique()

1160

'_id.$oid' is unique and not null and is hence the primary key.
barcode seems to have a few duplicates. Let's investigate further

In [11]:
# Looking at barcodes that are duplicated
brands[brands.duplicated(subset='barcode', keep=False)].sort_values('barcode')

barcode             category categoryCode                  name  \
467   511111004790               Baking          NaN                 alexa   
1071  511111004790  Condiments & Sauces          NaN       Bitten Dressing   
152   511111204923              Grocery          NaN                Brand1   
536   511111204923               Snacks          NaN             CHESTER'S   
20    511111305125                 Baby          NaN      Chris Image Test   
651   511111305125            Magazines          NaN  Rachael Ray Everyday   
129   511111504139            Beverages          NaN       Chris Brand XYZ   
299   511111504139              Grocery          NaN                  Pace   
9     511111504788               Baking          NaN                  test   
412   511111504788  Condiments & Sauces          NaN     The Pioneer Woman   
194   511111605058            Magazines          NaN       Health Magazine   
1012  511111605058                Dairy          NaN                Brand2   
1002  511111704140  Condiments & Sauces          NaN                 Prego   
1015  511111704140                  NaN          NaN       Diet Chris Cola   

     topBrand                  _id.$oid              cpg.$id.$oid cpg.$ref  \
467      True  5c409ab4cd244a3539b84162  55b62995e4b0d8e685c14213     Cogs   
1071      NaN  5cdacd63166eb33eb7ce0fa8  559c2234e4b06aca36af13c6     Cogs   
152      True  5c45f91b87ff3552f950f027  5c45f8b087ff3552f950f026     Cogs   
536       NaN  5d6027f46d5f3b23d1bc7906  5332f5fbe4b03c9a25efd0ba     Cogs   
20        NaN  5c4699f387ff3577e203ea29  55b62995e4b0d8e685c14213     Cogs   
651       NaN  5d642d65a3a018514994f42d  5d5d4fd16d5f3b23d1bc7905     Cogs   
129       NaN  5a7e0604e4b0aedb3b84afd3  55b62995e4b0d8e685c14213     Cogs   
299     False  5a8c33f3e4b07f0a2dac8943  5a734034e4b0d58f376be874     Cogs   
9         NaN  5c408e8bcd244a1fdb47aee7  59ba6f1ce4b092b29c167346     Cogs   
412       NaN  5ccb2ece166eb31bbbadccbe  559c2234e4b06aca36af13c6     Cogs   
194       NaN  5d6415d5a3a018514994f429  5d5d4fd16d5f3b23d1bc7905     Cogs   
1012     True  5c4637ba87ff35681e840d57  5c45f8b087ff3552f950f026     Cogs   
1002    False  5a8c344ae4b0ccf165fac9e5  5a734034e4b0d58f376be874     Cogs   
1015      NaN  5a7e0665e4b0aedb3b84afd4  55b62995e4b0d8e685c14213     Cogs   

          brandCode  
467           ALEXA  
1071         BITTEN  
152      0987654321  
536        CHESTERS  
20       CHRISIMAGE  
651    511111305125  
129        CHRISXYZ  
299            PACE  
9              TEST  
412   PIONEER WOMAN  
194    511111605058  
1012    09090909090  
1002          PREGO  
1015     DIETCHRIS2

We know that barcodes are unique to a specific product so they shouldn't be repeating. One explanation for the duplicates is that the superfluous ones are test codes that have been created but never deleted. We get to know this by looking at the 'name' column and noticing that names like "test", "Brand2", "Chris Image Test" etc are the duplicates. I would delete these from the dataset.

#### Checking receipts.json

In [12]:
receipts = jsonToDf('receipts')
receipts

bonusPointsEarned                            bonusPointsEarnedReason  \
0                 500.0  Receipt number 2 completed, bonus point schedu...   
1                 150.0  Receipt number 5 completed, bonus point schedu...   
2                   5.0                         All-receipts receipt bonus   
3                   5.0                         All-receipts receipt bonus   
4                   5.0                         All-receipts receipt bonus   
...                 ...                                                ...   
1114               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1115                NaN                                                NaN   
1116                NaN                                                NaN   
1117               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1118                NaN                                                NaN   

     pointsEarned  purchasedItemCount  \
0           500.0                 5.0   
1           150.0                 2.0   
2               5                 1.0   
3             5.0                 4.0   
4             5.0                 2.0   
...           ...                 ...   
1114         25.0                 2.0   
1115          NaN                 NaN   
1116          NaN                 NaN   
1117         25.0                 2.0   
1118          NaN                 NaN   

                                 rewardsReceiptItemList rewardsReceiptStatus  \
0     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2     [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
...                                                 ...                  ...   
1114  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1115                                                NaN            SUBMITTED   
1116                                                NaN            SUBMITTED   
1117  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1118                                                NaN            SUBMITTED   

     totalSpent                    userId                  _id.$oid  \
0         26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1         11.00  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   
2         10.00  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a   
3         28.00  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f   
4          1.00  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561   
...         ...                       ...                       ...   
1114      34.96  5fc961c3b8cfca11a077dd33  603cc0630a720fde100003e6   
1115        NaN  5fc961c3b8cfca11a077dd33  603d0b710a720fde1000042a   
1116        NaN  5fc961c3b8cfca11a077dd33  603cf5290a720fde10000413   
1117      34.96  5fc961c3b8cfca11a077dd33  603ce7100a7217c72c000405   
1118        NaN  5fc961c3b8cfca11a077dd33  603c4fea0a7217c72c000389   

      createDate.$date  dateScanned.$date  finishedDate.$date  \
0        1609687531000      1609687531000        1.609688e+12   
1        1609687483000      1609687483000        1.609687e+12   
2        1609687537000      1609687537000                 NaN   
3        1609687534000      1609687534000        1.609688e+12   
4        1609687506000      1609687506000        1.609688e+12   
...                ...                ...                 ...   
1114     1614594147000      1614594147000                 NaN   
1115     1614613361873      1614613361873                 NaN   
1116     1614607657664      1614607657664                 NaN   
1117     1614604048000      1614604048000                 NaN   
1118     1614565354962      1614565354962      

In [13]:
dateConvert(receipts, 'createDate.$date')
dateConvert(receipts, 'dateScanned.$date')
dateConvert(receipts, 'finishedDate.$date')
dateConvert(receipts, 'modifyDate.$date')
dateConvert(receipts, 'pointsAwardedDate.$date')
dateConvert(receipts, 'purchaseDate.$date')
receipts

bonusPointsEarned                            bonusPointsEarnedReason  \
0                 500.0  Receipt number 2 completed, bonus point schedu...   
1                 150.0  Receipt number 5 completed, bonus point schedu...   
2                   5.0                         All-receipts receipt bonus   
3                   5.0                         All-receipts receipt bonus   
4                   5.0                         All-receipts receipt bonus   
...                 ...                                                ...   
1114               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1115                NaN                                                NaN   
1116                NaN                                                NaN   
1117               25.0                        COMPLETE_NONPARTNER_RECEIPT   
1118                NaN                                                NaN   

     pointsEarned  purchasedItemCount  \
0           500.0                 5.0   
1           150.0                 2.0   
2               5                 1.0   
3             5.0                 4.0   
4             5.0                 2.0   
...           ...                 ...   
1114         25.0                 2.0   
1115          NaN                 NaN   
1116          NaN                 NaN   
1117         25.0                 2.0   
1118          NaN                 NaN   

                                 rewardsReceiptItemList rewardsReceiptStatus  \
0     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2     [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4     [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
...                                                 ...                  ...   
1114  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1115                                                NaN            SUBMITTED   
1116                                                NaN            SUBMITTED   
1117  [{'barcode': 'B076FJ92M4', 'description': 'mue...             REJECTED   
1118                                                NaN            SUBMITTED   

     totalSpent                    userId                  _id.$oid  \
0         26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1         11.00  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   
2         10.00  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a   
3         28.00  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f   
4          1.00  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561   
...         ...                       ...                       ...   
1114      34.96  5fc961c3b8cfca11a077dd33  603cc0630a720fde100003e6   
1115        NaN  5fc961c3b8cfca11a077dd33  603d0b710a720fde1000042a   
1116        NaN  5fc961c3b8cfca11a077dd33  603cf5290a720fde10000413   
1117      34.96  5fc961c3b8cfca11a077dd33  603ce7100a7217c72c000405   
1118        NaN  5fc961c3b8cfca11a077dd33  603c4fea0a7217c72c000389   

            createDate.$date       dateScanned.$date  finishedDate.$date  \
0    2021-01-03 15:25:31.000 2021-01-03 15:25:31.000 2021-01-03 15:25:31   
1    2021-01-03 15:24:43.000 2021-01-03 15:24:43.000 2021-01-03 15:24:43   
2    2021-01-03 15:25:37.000 2021-01-03 15:25:37.000                 NaT   
3    2021-01-03 15:25:34.000 2021-01-03 15:25:34.000 2021-01-03 15:25:34   
4    2021-01-03 15:25:06.000 2021-01-03 15:25:06.000 2021-01-03 15:25:11   
...                      ...                     ...                 ...   
1114 2021-03-01 10:22:27.000 2021-03-01 10:22:27.000                 NaT   
1115 2021-03-01 15:42:41.873 2021-03-01 15:42:41.873                 NaT   
1116 2021-03-01 14:07:37.664 2021-03-01 14:07:37.664                 NaT   
11

In [14]:
# Looking at the '_id.$oid' column to check if this should be the primary key for the receipts table
receipts['_id.$oid'].nunique()

1119

In [15]:
receipts['rewardsReceiptItemList']

0       [{'barcode': '4011', 'description': 'ITEM NOT ...
1       [{'barcode': '4011', 'description': 'ITEM NOT ...
2       [{'needsFetchReview': False, 'partnerItemId': ...
3       [{'barcode': '4011', 'description': 'ITEM NOT ...
4       [{'barcode': '4011', 'description': 'ITEM NOT ...
                              ...                        
1114    [{'barcode': 'B076FJ92M4', 'description': 'mue...
1115                                                  NaN
1116                                                  NaN
1117    [{'barcode': 'B076FJ92M4', 'description': 'mue...
1118                                                  NaN
Name: rewardsReceiptItemList, Length: 1119, dtype: object

Each row is a list of key value pairs having the same receipt_id. 'rewardsReceiptItemList' will be modeled as a separate table called 'items', as shown in the ERD.